CONTRIBUTING.md

```diff
@@ -51,26 +51,125 @@ You can file new issues by filling out our [new issue form](https://github.com/N
 ### Forking
 Stark development is done using a forking model with Pull Requests (PRs), so the very first thing you should do is create your fork: https://help.github.com/articles/fork-a-repo/
 
+We heavily recommend using the "GitFlow" workflow (see workflow section below) with feature branches.
+
 ### Integrating changes to your fork
 Once you're found what you want to contribute to Stark, then:
 * Create a feature branch in your fork: `git checkout -b my-new-branch master`
-* Configure the upstream: https://help.github.com/articles/configuring-a-remote-for-a-fork/
+* Configure the upstream
+  * `git remote add upstream https://github.com/NationalBankBelgium/stark.git`
+  * reference: https://help.github.com/articles/configuring-a-remote-for-a-fork/
 
 From then on, you may work on your feature at your own rhythm and commit/push to your fork.
 
-Meanwhile, don't forget to keep your fork up to date: https://help.github.com/articles/syncing-a-fork/
+Don't forget to write test cases (or adapt existing ones) for your changes! 
+
+### Keeping in sync
+While you're busy developing your feature and before you propose them, make sure that your fork is up to date with the upstream.
+
+First, download the latest commits:
+* `git fetch upstream`
+* or `git remote update -p`
+
+Then, integrate those changes to your fork (whatever branch you're working on).
+
+First try a fast-forward merge: `git merge --ff-only @{u}`
+  * that command tells git to merge the upstream branch ONLY if you local branch can be "fast forwarded" to the upstream branch (i.e., if it hasn't diverged)
+
+If the fast-forward merge fails, then you'll have to rebase with the upstream (i.e., align): `git rebase -p @{u}`
+  * the `-p` options tells git to preserve merges. This prevents git from linearizing the commits being rebased
+
+Once done, make sure the history looks like what you expect: `git log --graph --oneline --decorate --date-order --color --boundary @{u}`
+
 Certainly so before creating a Pull Request (PR). If you don't do it then we'll request it anyways.
 
-Don't forget to write test cases (or adapt existing ones) for your changes! 
+References
+  * https://stackoverflow.com/questions/6406762/why-am-i-merging-remote-tracking-branch-origin-develop-into-develop
+  * https://help.github.com/articles/syncing-a-fork/
 
 ### Proposing your changes by submitting a Pull Request (PR)
-Before you propose your changes, make sure that your fork is up to date with the upstream and that the whole test suite passes.
-Also, try and rebase / squash your commits in order to keep the history clean.
+Before you propose your changes, make sure that your fork is up to date with the upstream (see previous section) and that the whole test suite passes.
+
+At this point, you'll probably want to do a rebase in order to keep your commits to a minimum amount, while making sure that the contents of those commits make sense together.
```

Comment:
Mmm I don't fully understand this line... I thought rebase was used to keep the linearity of the history and squash to condense several commits into one. Or am I missing something?

Perhaps it could be useful to also talk about squashing commits in this section...

Reply:
Yes that's what I meant, wasn't clear though, I'll try to rephrase